pkg/compactor/block_upload_test.go

```diff
@@ -0,0 +1,735 @@
+			},
+			verifyUpload: func(t *testing.T, bkt *bucket.ClientMock) {
+				verifyUpload(t, bkt, map[string]string{})
+			},
+		},
+		{
+			name:            "valid request without compactor shard ID label",
+			tenantID:        tenantID,
+			blockID:         blockID,
+			setUpBucketMock: setUpUpload,
+			meta: &metadata.Meta{
+				BlockMeta: tsdb.BlockMeta{
+					ULID:    bULID,
+					Version: metadata.TSDBVersion1,
+					MinTime: now - 1000,
+					MaxTime: now,
+				},
+				Thanos: metadata.Thanos{
+					Files: []metadata.File{
+						{
+							RelPath: block.MetaFilename,
+						},
+						{
+							RelPath:   "index",
+							SizeBytes: 1,
+						},
+						{
+							RelPath:   "chunks/000001",
+							SizeBytes: 1024,
+						},
+					},
+				},
+			},
+			verifyUpload: func(t *testing.T, bkt *bucket.ClientMock) {
+				verifyUpload(t, bkt, nil)
+			},
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			var bkt bucket.ClientMock
+			if tc.setUpBucketMock != nil {
+				tc.setUpBucketMock(&bkt)
+			}
+
+			cfgProvider := newMockConfigProvider()
+			cfgProvider.userRetentionPeriods[tenantID] = tc.retention
+			c := &MultitenantCompactor{
+				logger:       log.NewNopLogger(),
+				bucketClient: &bkt,
+				cfgProvider:  cfgProvider,
+			}
+			var rdr io.Reader
+			if tc.body != "" {
+				rdr = bytes.NewReader([]byte(tc.body))
+			} else if tc.meta != nil {
+				buf := bytes.NewBuffer(nil)
+				require.NoError(t, json.NewEncoder(buf).Encode(tc.meta))
+				rdr = buf
+			}
+			r := httptest.NewRequest(http.MethodPost, fmt.Sprintf("/api/v1/upload/block/%s", tc.blockID), rdr)
+			if tc.tenantID != "" {
+				r = r.WithContext(user.InjectOrgID(r.Context(), tc.tenantID))
+			}
+			if tc.blockID != "" {
+				r = mux.SetURLVars(r, map[string]string{"block": tc.blockID})
+			}
+			w := httptest.NewRecorder()
+			c.HandleBlockUpload(w, r)
+
+			resp := w.Result()
+			body, err := io.ReadAll(resp.Body)
+			require.NoError(t, err)
+
+			switch {
+			case tc.expInternalServerError:
+				assert.Equal(t, http.StatusInternalServerError, resp.StatusCode)
+				assert.Equal(t, "internal server error\n", string(body))
+			case tc.expBadRequest != "":
+				assert.Equal(t, http.StatusBadRequest, resp.StatusCode)
+				assert.Equal(t, fmt.Sprintf("%s\n", tc.expBadRequest), string(body))
+			case tc.expConflict != "":
+				assert.Equal(t, http.StatusConflict, resp.StatusCode)
+				assert.Equal(t, fmt.Sprintf("%s\n", tc.expConflict), string(body))
+			default:
+				assert.Equal(t, http.StatusOK, resp.StatusCode)
+				assert.Empty(t, string(body))
+			}
+
+			bkt.AssertExpectations(t)
+
+			if tc.verifyUpload != nil {
+				tc.verifyUpload(t, &bkt)
+			}
+		})
+	}
+}
+
+// Test MultitenantCompactor.UploadBlockFile.
+func TestMultitenantCompactor_UploadBlockFile(t *testing.T) {
+	const tenantID = "test"
+	const blockID = "01G3FZ0JWJYJC0ZM6Y9778P6KD"
+	uploadingMetaPath := path.Join(tenantID, blockID, fmt.Sprintf("uploading-%s", block.MetaFilename))
```

Comment:
@pstibrany actually, I avoided referring to `tmpMetaFilename` in tests intentionally, because I want tests to guard against an erroneous change to `tmpMetaFilename` in the implementation.